runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/GeomancyManager.java

```diff
@@ -0,0 +1,238 @@
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.timetracking.farming;
+
+import java.time.Instant;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import net.runelite.api.Client;
+import net.runelite.api.ItemID;
+import net.runelite.api.events.ScriptPreFired;
+import net.runelite.client.Notifier;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.eventbus.Subscribe;
+import net.runelite.client.plugins.timetracking.TimeTrackingConfig;
+
+@Slf4j
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE, onConstructor = @__({@Inject}))
+@Singleton
+public class GeomancyManager
+{
+	private final Client client;
+	private final ConfigManager configManager;
+	private final FarmingWorld farmingWorld;
+	private final TimeTrackingConfig config;
+	private final PaymentTracker paymentTracker;
+	private final CompostTracker compostTracker;
+	private final FarmingTracker farmingTracker;
+	private final GeomancyMaps geomancyMaps = new GeomancyMaps();
+
+	@Inject
+	private Notifier notifier;
+	private boolean isDirty;
+	private boolean enableVerboseLogging = false;
+
+	@Subscribe
+	public void onScriptPreFired(ScriptPreFired scriptPreFired)
+	{
+		if (scriptPreFired.getScriptId() == 1119)
+		{
+			//args[0] is script ID
+			Object[] args = scriptPreFired.getScriptEvent().getArguments();
+			int jagexPatchID = (int) args[1];
+			int cropItemID = (int) args[2];
+			int coords = (int) args[3];
+			int flags = (int) args[4];
+
+			GeomancyData data = new GeomancyData(coords);
+
+			if (enableVerboseLogging)
+			{
+				LogPatchInfo(jagexPatchID, cropItemID, data.minimumYield, data.isWatered, data.currentStage, data.maxStage, flags);
+			}
+
+			FarmingPatch fp = farmingWorld.getPatchByJagexID(jagexPatchID);
+			if (fp == null)
+			{
+				log.error("Couldn't find farming patch to match Jagex patch ID " + jagexPatchID);
```

Comment:
warn + use format specifiers and not string concatenation

Reply:
addressed in 05a67583d0f090b5518c0c1ad2ab59c2bff6c75d